Gradio Guides: Creating CryptoPunks with GANs

NimaBoscarino opened the pull request:
# Description
This PR adds a step-by-step guide for making a GAN-powered image generator, complete with a Gradio interface. The learning outcomes of this guide are:

* Using an image output component
* Adding examples to a Gradio demo
* Exposure to GANs at a high-level, and seeing a Generator in action

# Checklist:

- [X] I have performed a self-review of my own code
- [X] My code follows the style guidelines of this project
- [X] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

aliabd commented on guides/generate_punks_with_gan.md:
````diff
@@ -0,0 +1,164 @@
+    outputs="image",
+).launch()
+```
+
````
Suggestion: this would be a good place to embed an interface so readers can immediately interact with what's been built so far (before the second slider). To do so you'll need to create a version of the space from this gradio config

aliabd commented on guides/generate_punks_with_gan.md:
```diff
@@ -0,0 +1,164 @@
+        gr.inputs.Slider(label='Seed', minimum=0, maximum=1000, default=42),
+        gr.inputs.Slider(label='Number of Punks', minimum=4, maximum=64, step=1, default=10), # Adding another slider!
+    ],
+    # ...etc.
```
Given that there is not much left in the `...etc.` I would include the full code. Code snippets are easy to copy on guides so when applicable it's better to make them immediately useable.

aliabd commented on guides/generate_punks_with_gan.md:
````diff
@@ -0,0 +1,164 @@
+    return 'punks.png'
+```
+
+When you start up your Gradio app again, you should see a second slider that'll let you control the number of punks!
````
Same comment as before about embedding another interface, now with the new slider, but without the polishing portion

aliabd replied:
Small comment about changing `start up your gradio app` to `relaunch your interface`

aliabd commented on guides/generate_punks_with_gan.md:
```diff
@@ -0,0 +1,164 @@
+
+The `examples` parameter takes a list of lists, where each item in the sublists is ordered in the same order that we've listed the `inputs`. So in our case, `[seed, num_punks]`. Give it a try!
+
+You can also try adding a `title`, `description`, and `article` to the `gr.Interface`. Each of those parameters accepts a string, so try it out and see what happens 👀 `article` will also accept HTML, which we'll explore in a future guide!
```
Would be cool to now include a full code snippet including all the things in https://huggingface.co/spaces/NimaBoscarino/cryptopunks/blob/main/app.py which we have now recreated, as well as re-embedding the full interface. If you think its repetitive then not necessary, but creates a better reader experience in my opinion.

aliabd replied:
article has been explored in a [previous guide](http://localhost:8000/adding_rich_descriptions_to_your_demo/)! feel free to link :)

aliabd commented on guides/generate_punks_with_gan.md:
```diff
@@ -0,0 +1,164 @@
+
+You can also try adding a `title`, `description`, and `article` to the `gr.Interface`. Each of those parameters accepts a string, so try it out and see what happens 👀 `article` will also accept HTML, which we'll explore in a future guide!
+
+----------
```
FYI this gets rendered as slightly more space. not sure if that's what's intended.
<img width="931" alt="Screen Shot 2022-04-19 at 12 10 55 AM" src="https://user-images.githubusercontent.com/9021060/163885619-3fa654d7-8619-474a-a880-bca5a6a2852a.png">

NimaBoscarino replied:
Yup! I just copied the format used in some of the other guides (e.g. https://github.com/gradio-app/gradio/blob/main/guides/image_classification_in_pytorch.md?plain=1#L87)

aliabd commented:
Terrific job @NimaBoscarino! This is one of the more unique guides we've had written. I left some comments but overall super well written :)

NimaBoscarino commented on guides/generate_punks_with_gan.md:
```diff
@@ -0,0 +1,164 @@
+# Create Your Own Friends with a GAN
+
+related_spaces: https://huggingface.co/spaces/NimaBoscarino/cryptopunks, https://huggingface.co/nateraw/cryptopunks-gan
```
Ah good catch, that's the one!

NimaBoscarino commented:
Wooooo glad you like it, and thanks for the review!! I think I addressed all the suggestions, let me know if there's anything I missed! I haven't been able to run the website UI on my laptop yet to verify how the guide renders, but I hope it looks okay.

aliabd commented:
@NimaBoscarino your changes look good, but [this space](https://huggingface.co/spaces/NimaBoscarino/cryptopunks) is broken because of a runtime error, which is causing the embedding to fail. 

Looks like its because the examples don't include the second slider. (also, can you include the examples in the final code snippet in the guide?)

NimaBoscarino commented:
Ah that's embarrassing, sorry about that @aliabd! 😅  I fixed up the space and made sure to include the examples in the final snippet.

aliabd commented:
Looks great! :) will merge. Great work!
